ZEP008-hot-backup.md

```diff
@@ -0,0 +1,619 @@
+
+![message reordering](ZEP008/message-reordering.png)
+
+In this scenario, reordered messages does not lead to an inconsistent checkpoint.
+In $CP-2$, neither of the messages are correlated.
+
+![message reordering](ZEP008/message-reordering-2.png)
+
+Here, in $CP-2$, message 2 is published and correlated, but message 1 is not correlated. This is ok, because this can also happen in a normal execution. When the cluster restore from this state, message 1 correlation command will be re-send.
+
+### Detailed Backup Process in Zeebe
+
+Taking backup of a partition can take a long time. We do not want to block the StreamProcessor during this period. To prevent blocking, we should allow the backup to be taken concurrently to the StreamProcessor.
+When the backup is taken concurrently, it leads to other challenges. Snapshot is also taken concurrently, and a concurrent snapshot can lead to log compaction. So before the backup is taken, a new snapshot might have been taken resulting in deleting the old snapshot and compacting the log. The new snapshot might be already at a position further to the checkpointPosition rendering the backup to be invalid.
+The challenge here is to allow asynchronous backup while preventing concurrent snapshots and compaction from making the backup invalid.
+
+We extend the high-level process described in [Guide-level explanation](internal-bakcup-process) to allow asynchronous backup as follows.
+
+The components involved in Backup.
+* A Coordinator
+* BackupStore
+In a partition:
+	- StreamProcessor
+	- SnapshotDirector
+	- BackupActor
+	- SnapshotStoreActor
+
+###### Coordinator
+A backup coordinator is external to all partitions. The gateway could act as a coordinator.
+
+To take a backup `backupId`, coordinator send the command `checkpoint backupId` to all partitions.
+The coordinator receives an acknowledgment from the partitions when the partitions have started taking the backup (when the StreamProcessor has processed the command). Note that it does not wait until the backup has completed, because it can take a long time.
+
+To monitor the status of a backup, the coordinator sends a request to a partition.
+The partition can then check the status of its backup and sends a response.
+The coordinator then aggregates the status from all partitions.
+
+###### BackupStore
+BackupStore is a remote storage where each partition can backup it's data. Backup[N] refers to a backup with id N.
+
+Backup[N][p][b].files = contains backup N of partition `p` started by broker `b`
+Backup[N][p][b].status = doesNotExist | ongoing | completed | failed
+
+Backup[N].status = completed if for all partitions p, Backup[N][p].status= completed.
+Backup[N].status= failed if for atleast one partition p, Backup[N][p].status=failed.
+Backup[N].status = ongoing if for atleast one partition p, Backup[N][p].status=ongoing and Backup[N].status != failed.
+
+###### BackupActor
+
+On receiving the command `takeBackup checkpointId position` :
```

Comment:
Yes. It is the position of the command that triggers the checkpoint. It could be the position of the "checkpoint" record or a remote command that triggered the checkpoint.